cmd/private-org-sync/main.go

```diff
@@ -0,0 +1,469 @@
+const unshallow = maxExpDepth + 1
+
+func makeFetch(logger *logrus.Entry, repoDir string, git gitFunc, remote, branch string, expDepth int) func() error {
+	return func() error {
+		fetch := []string{"fetch", remote, branch}
+
+		depthArg := "full fetch" // no depth arg is used when doing a full fetch
+		if expDepth != fullFetch {
+			depthArg = "--unshallow"
+			if expDepth < unshallow {
+				depthArg = fmt.Sprintf("--depth=%d", int(math.Exp2(float64(expDepth))))
+			}
+
+			fetch = append(fetch, depthArg)
+		}
+
+		logger.Infof("Fetching from source (%s)", depthArg)
+		if out, exitCode, err := git(logger, repoDir, fetch...); err != nil || exitCode != 0 {
+			logger.WithError(err).WithField("exit-code", exitCode).WithField("output", out).Error("Failed to fetch from source")
+			return fmt.Errorf("failed to fetch from source")
+		}
+		return nil
+	}
+}
+
+func maybeTooShallow(pushOutput string) bool {
+	patterns := []string{
+		"shallow update not allowed",
+		"Updates were rejected because the remote contains work that you do",
+	}
+	for _, item := range patterns {
+		if strings.Contains(pushOutput, item) {
+			return true
+		}
+	}
+	return false
+}
+
+// location specifies a GitHub repository branch used as a source or destination
+type location struct {
+	org, repo, branch string
+}
+
+func (l location) String() string {
+	return fmt.Sprintf("%s/%s@%s", l.org, l.repo, l.branch)
+}
+
+// makeGitDir creates a directory for a local git repo used for fetching content
+// from the given location and pushing it to any other git repo
+func (g gitSyncer) makeGitDir(org, repo string) (string, error) {
+	repoDir := filepath.Join(g.root, org, repo)
+	err := os.MkdirAll(repoDir, 0755)
+	return repoDir, err
+}
+
+// mirror syncs content from source location to destination one, using a local
+// repository in the given path. The `repoDir` directory must exist and be
+// either empty, or previously used in a `mirror()` call. If it is empty,
+// a bare git repository will be initialized in it. The git content from
+// the `src` location will be fetched to this local repository and then
+// pushed to the `dst` location. Multiple `mirror` calls over the same `repoDir`
+// will reuse the content fetched in previous calls, acting like a cache.
+func (g gitSyncer) mirror(repoDir string, src, dst location) error {
+	mirrorFields := logrus.Fields{
+		"source":      src.String(),
+		"destination": dst.String(),
+		"local-repo":  repoDir,
+	}
+	logger := g.logger.WithFields(mirrorFields)
+	logger.Info("Syncing content between locations")
+
+	logger.Debug("Initializing git repository")
+	if _, exitCode, err := g.git(logger, repoDir, "init", "--bare"); err != nil || exitCode != 0 {
+		logger.WithField("exit-code", exitCode).WithError(err).Error("Failed to initialize local git directory")
+		return fmt.Errorf("failed to initialize local git directory")
+	}
+
+	srcRemote := fmt.Sprintf("%s-%s", src.org, src.repo)
```

Comment:
Why do you need to configure the `git` to have remotes? `git fetch`, `git ls-remote` and `git push` all can just take the remote as an arg

Reply:
Good question, looks like I misplaced the comment with the explanation (I swear I wrote it but it got lost in the revisions). Without the remote, the local repo does not serve as a cache:

```console
$ git init --bare
Initialized empty Git repository in /home/afri/Temporary/bare-origin/
$ time git fetch https://github.com/openshift/origin.git --depth 2 master
remote: Enumerating objects: 67292, done.
remote: Counting objects: 100% (67292/67292), done.
remote: Compressing objects: 100% (44239/44239), done.
remote: Total 67292 (delta 22115), reused 44537 (delta 17109), pack-reused 0
Receiving objects: 100% (67292/67292), 114.47 MiB | 10.84 MiB/s, done.
Resolving deltas: 100% (22115/22115), done.
From https://github.com/openshift/origin
 * branch                master     -> FETCH_HEAD
7.18user 1.32system 0:25.63elapsed 33%CPU (0avgtext+0avgdata 88496maxresident)k
0inputs+238144outputs (0major+30789minor)pagefaults 0swaps
$ time git fetch https://github.com/openshift/origin.git --depth 2 master
remote: Enumerating objects: 67292, done.
remote: Counting objects: 100% (67292/67292), done.
remote: Compressing objects: 100% (44239/44239), done.
remote: Total 67292 (delta 22115), reused 44537 (delta 17109), pack-reused 0
Receiving objects: 100% (67292/67292), 114.47 MiB | 10.80 MiB/s, done.
Resolving deltas: 100% (22115/22115), done.
From https://github.com/openshift/origin
 * branch                master     -> FETCH_HEAD
8.88user 1.33system 0:26.67elapsed 38%CPU (0avgtext+0avgdata 332716maxresident)k
0inputs+238144outputs (0major+81533minor)pagefaults 0swaps
$ rm -rf *; git init --bare
Initialized empty Git repository in /home/afri/Temporary/bare-origin/
$ git remote add oo https://github.com/openshift/origin.git
$ time git fetch oo --depth 2 master
remote: Enumerating objects: 67292, done.
remote: Counting objects: 100% (67292/67292), done.
remote: Compressing objects: 100% (44239/44239), done.
remote: Total 67292 (delta 22115), reused 44537 (delta 17109), pack-reused 0
Receiving objects: 100% (67292/67292), 114.47 MiB | 10.67 MiB/s, done.
Resolving deltas: 100% (22115/22115), done.
From https://github.com/openshift/origin
 * branch                master     -> FETCH_HEAD
 * [new branch]          master     -> oo/master
7.20user 1.45system 0:27.39elapsed 31%CPU (0avgtext+0avgdata 88852maxresident)k
16inputs+238152outputs (0major+29544minor)pagefaults 0swaps
$ time git fetch oo --depth 2 master
remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
From https://github.com/openshift/origin
 * branch                master     -> FETCH_HEAD
0.13user 0.04system 0:01.88elapsed 9%CPU (0avgtext+0avgdata 88112maxresident)k
0inputs+16outputs (0major+9670minor)pagefaults 0swaps
```

But you are right, this definitely needs explanation.

Reply:
Added an explanatory comment to the code.